### PR TITLE
[SYCL][NFC] Add SYCLPropagateAspectsUsage pass

### PR DESCRIPTION
**Pull request by AlexeySachkov**
Added a pass which is a part of optional kernel features design: it uses
information provieded by FE & Headers about aspects used in device code
to propagate it through the call graph to mark all kernels and functions
with list of aspects they use.

Co-authored-by: Maksim Sabianin <maksim.sabianin@intel.com>

### CODE REVIEW

**Comment by AlexeySachkov** [comment]
This is a partial re-submit of #5348. I intentionally reduced the scope of the patch to simplify review and avoid some stuff which caused a lot of discussions and requires more work to be properly acceptable.

The PR adds a pass in accordance with optional kernel features [design document](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/OptionalDeviceFeatures.md). For now, the only thing the pass can do is to collect optional kernel features usages and propagate info about them through a call graph, adding necessary metadata to functions and kernels.

The pass is not yet included into a pipeline, that will be done in follow-up PRs together with consumers of metadata emitted by the pass

**Comment by AlexeySachkov** [comment]
@asudarsa, could you please take a look?

**Review by KornevNikita — APPROVED** [review]

File: llvm/include/llvm/SYCLLowerIR/PropagateAspectsUsage.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#include "llvm/IR/PassManager.h"
+
+namespace llvm {
+
+class PropagateAspectsUsagePass
+    : public PassInfoMixin<PropagateAspectsUsagePass> {
+public:
+  PreservedAnalyses run(Module &M, ModuleAnalysisManager &);
+};
+
```
Comment (asudarsa) [inline llvm/include/llvm/SYCLLowerIR/PropagateAspectsUsage.h:26]: Are we not creating 'legacy' versions anymore?

File: llvm/include/llvm/SYCLLowerIR/PropagateAspectsUsage.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
+//===---- PropagateAspectsUsage.h - PropagateAspectsUsage Pass ------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Pass propagates metadata corresponding to usage of optional device
```
Comment (asudarsa) [inline llvm/include/llvm/SYCLLowerIR/PropagateAspectsUsage.h:9]: Rephrase: ... optional device features through the callgraph.

File: llvm/lib/SYCLLowerIR/PropagateAspectsUsage.cpp [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Pass propagates metadata corresponding to usage of optional device
+// features.
+//
+// The pass consists of a four main steps:
```
Comment (asudarsa) [inline llvm/lib/SYCLLowerIR/PropagateAspectsUsage.cpp:12]: NIT: consists of four main steps

File: llvm/lib/Passes/PassRegistry.def [hunk 4]
```diff
@@ -136,6 +136,7 @@ MODULE_PASS("SPIRITTAnnotations", SPIRITTAnnotationsPass())
 MODULE_PASS("deadargelim-sycl", DeadArgumentEliminationSYCLPass())
 MODULE_PASS("sycllowerwglocalmemory", SYCLLowerWGLocalMemoryPass())
 MODULE_PASS("lower-esimd-kernel-attrs", SYCLFixupESIMDKernelWrapperMDPass())
+MODULE_PASS("sycl-propagate-aspects-usage", PropagateAspectsUsagePass())
```
Comment (asudarsa) [inline llvm/lib/Passes/PassRegistry.def:139]: May be rename Pass with SYCL prefix for uniformity?

File: llvm/lib/SYCLLowerIR/PropagateAspectsUsage.cpp [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+    FunctionToAspects[&F].insert(Aspects.begin(), Aspects.end());
+
+    if (const auto *CI = dyn_cast<CallInst>(&I)) {
+      if (!CI->isIndirectCall() && CI->getCalledFunction())
+        CG[&F].insert(CI->getCalledFunction());
+    }
+  }
+}
+
+// Return true if the function is a SPIRV or SYCL builtin, e.g.
```
Comment (asudarsa) [inline llvm/lib/SYCLLowerIR/PropagateAspectsUsage.cpp:278]: May be a comment here saying this (and the next function) code is shared with ModuleSplitter.cpp and mention that the two need to be in sync?

**Review by asudarsa — COMMENTED** [review]
Looks good overall. A few nitpicks. Also, I wanted to check where this pass will be called eventually.
Thanks.

File: llvm/include/llvm/SYCLLowerIR/PropagateAspectsUsage.h (hunk 1 above)
Comment (AlexeySachkov) [reply llvm/include/llvm/SYCLLowerIR/PropagateAspectsUsage.h:26]: This pass is supposed to be used from `clang/lib/CodeGen/BackendUtil.cpp`, i.e. from Clang middle-end passes, which use new pass manager without possibility to switch to legacy one

File: llvm/include/llvm/SYCLLowerIR/PropagateAspectsUsage.h (hunk 2 above)
Comment (AlexeySachkov) [reply llvm/include/llvm/SYCLLowerIR/PropagateAspectsUsage.h]: Change in 96139d671baf3ea32672932b73d2dc5c0a56663b

File: llvm/lib/SYCLLowerIR/PropagateAspectsUsage.cpp (hunk 3 above)
Comment (AlexeySachkov) [reply llvm/lib/SYCLLowerIR/PropagateAspectsUsage.cpp]: Applied in 96139d671baf3ea32672932b73d2dc5c0a56663b

File: llvm/lib/Passes/PassRegistry.def (hunk 4 above)
Comment (AlexeySachkov) [reply llvm/lib/Passes/PassRegistry.def]: Done in 96139d671baf3ea32672932b73d2dc5c0a56663b

File: llvm/lib/SYCLLowerIR/PropagateAspectsUsage.cpp (hunk 5 above)
Comment (AlexeySachkov) [reply llvm/lib/SYCLLowerIR/PropagateAspectsUsage.cpp]: Added such comment in 96139d671baf3ea32672932b73d2dc5c0a56663b

File: llvm/include/llvm/SYCLLowerIR/SYCLPropagateAspectsUsage.h [hunk 6]
```diff
@@ -0,0 +1,28 @@
+//===---- PropagateAspectsUsage.h - PropagateAspectsUsage Pass ------------===//
```
Comment (asudarsa) [inline llvm/include/llvm/SYCLLowerIR/SYCLPropagateAspectsUsage.h:1]: Please change name here as well.

**Review by asudarsa — CHANGES_REQUESTED** [review]
Just one correction. Otherwise, looks good.

Thanks

File: llvm/include/llvm/SYCLLowerIR/SYCLPropagateAspectsUsage.h (hunk 6 above)
Comment (AlexeySachkov) [reply llvm/include/llvm/SYCLLowerIR/SYCLPropagateAspectsUsage.h]: Fixed in 3b4d9db4727ce6c7b516946df8fbfa680fd65a25

**Comment by AlexeySachkov** [comment]
HIP XPASSES are unrelated and they were resolved in intel/llvm-test-suite#1236

**Review by vmaksimo — APPROVED** [review]

**Comment by bader** [comment]
@AlexeySachkov, this PR breaks `check-llvm` for the configuration with enabled opaque pointer support. Could you take a look and fix tests to work with opaque pointers, please?

**Comment by AlexeySachkov** [comment]
> @AlexeySachkov, this PR breaks `check-llvm` for the configuration with enabled opaque pointer support. Could you take a look and fix tests to work with opaque pointers, please?

@bader, I've published a fix in #6767

**Pull request merged**